odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,497 @@
+
+        Examples
+        --------
+        Deform a 1D function ``template`` that has value ``1`` at 3:rd sample
+        point using a displacement field ``displacement_field`` that has
+        value ``-0.2`` in the 4:th sample point. The outcome should be a
+        1D function that has values ``1`` at 3:rd and 4:th sample points.
+
+        >>> import odl
+        >>> space = odl.uniform_discr(0, 1, 5)
+        >>> template = space.element([0, 0, 1, 0, 0])
+        >>> op = LinDeforFixedTempOp(template)
+        >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+        >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+        >>> op(displacement_field)
+        uniform_discr(0.0, 1.0, 5).element([0.0, 0.0, 1.0, 1.0, 0.0])
+        """
+        assert isinstance(template.space, odl.DiscreteLp)
+
+        domain_space = odl.ProductSpace(template.space, template.space.ndim)
+        range_space = template.space
+
+        Operator.__init__(self, domain_space, range_space, linear=False)
+
+        self.template = template
+
+    def _call(self, displacement):
+        """Implementation of ``self(displacement)``.
+
+        Parameters
+        ----------
+        displacement : `ProductSpaceVector`
+            The displacement field used in the linearized deformation.
+        """
+
+        return linear_deform(self.template, displacement)
+
+    def derivative(self, displacement):
+        """Gateaux derivative of the operator in ``displacement``.
+
+        Parameters
+        ----------
+        displacement: `ProductSpaceVector`
+            The point that the Gateaux derivative need to be computed at.
+
+        Returns
+        -------
+        deriv_op : `Operator`
+            The derivative evaluated at ``displacement``
+        """
+        deriv_op = LinDeforFixedTempDeriv(self.template, displacement)
+        return deriv_op
+
+
+class LinDeforFixedTempDeriv(Operator):
+    """Derivative of the fixed template linearized deformation operator.
+
+    This operator computes the Gateaux derivative of the fixed
+    template linearized deformation operator for the fixed
+    template ``I`` at given displacement ``v``:
+
+        u --> grad(I)(Id + v).T u
+
+    where the ``u`` is a given vector field and
+    the ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement field:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain of the target.
+    In diffeomorphic sense, the domain of ``I`` is the same
+    as the target.
+    """
+
+    def __init__(self, template, displacement):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector`
+            Fixed template deformed by the vector field.
+        displacement: `ProductSpaceVector`
+            The point that the Gateaux derivative needs to be computed at.
```

Comment:
In a mathematical sense, you are perfectly right. But in code, we don't have a way of differentiating between different notions of derivative. In the current case, they are in fact equivalent because the Gateaux derivative is linear and bounded, so it is also the Frechet derivative. Thus, writing "Gateaux" does not add any information. 

The only relevant scenario I can think of is that we have an operator which is Gateaux differentiable but not Frechet differentiable, because the derivative is not linear. Then one could still implement `derivative` and set `linear=False`. In this case, and only in this case, I would agree that there is a need for documentation to answer the question "why is the derivative non-linear?".
In any other case, there is no need to write anything -- by giving `linear=True` and by not being able to check boundedness, we assume that the derivative is in the Frechet sense. Which is fine in 99.9 % of the use cases.